# Exploratory Data Analysis - Bank Customer Attrition

## Objectives

Explore the Bank Customer Attrition dataset to investiagte and carry out the following:

* Identify Outliers
* Explore feature engineering
* Descriptive Analysis
* Statistical Analysis

## Inputs

* data/inputs/cleaned_bank_data.csv

## Outputs

* INSERT CSV NAME HERE

___________________

In [1]:
# Import necessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [12]:
# Load cleaned dataset
df = pd.read_csv('../data/inputs/cleaned_bank_data.csv')

# Display first row of DataFrame
df.head(1)

Unnamed: 0  RowNumber  CustomerId  CreditScore Geography  Gender  Age  \
0           0          1    15598695          619    France  Female   42   

   Tenure  Balance  NumOfProducts  HasCrCard  IsActiveMember  EstimatedSalary  \
0       2      0.0              1          1               1        101348.88   

   Exited  Complain  Satisfaction Score Card Type  Point Earned  
0       1         1                   2   DIAMOND           464

_____________________

## Statistical Analysis

Explore the basic statistics from the cleaned datatset.

In [21]:
# Display basic statistics of the DataFrame
categorical_stats = df.describe(include='object')
display(categorical_stats)

Geography Gender Card Type
count      10000  10000     10000
unique         3      2         4
top       France   Male   DIAMOND
freq        5014   5457      2507

**Observations:**

* **Geography** - There are three unique entries within the column, which reveals the dataset based on customers across 3 countries. With the most customers living in France, as French customers make up 5014 of the 10000 entries. 

* **Gender** - The initial exploration reveals there are a greater number of male customers than female customers. The male customer population makes up 5457 of the 10000 entries.

* **Card Type** - There are 4 differnet card types, with DIAMOND being the most popular among all customers. The DIAMOND card holder population makes up 2507 of the 10000 entries. 

Investiagte the mean, median, standard deviation and range of the dataset.

In [23]:
# Perform basic statistical analysis
summary_stats = df.describe()
styled_summary_stats = summary_stats.style.background_gradient(cmap='Blues')
display(styled_summary_stats)

**Analysis Summary**:

This summary is based on the numeric columns relevant to the 3 hypotheses for the project.

**CreditScore:**

* **Mean** = 650.52

* **Median** = 652

* **Standard Deviation** = 96.65. This value suggest there is a great amount of variation between credit scores, which are both significantly above and below the mean credit score. 

* **Range** = 350 - 850. The range reflects the distirbution of credit scores within the dataset.


**NumOfProducts:**

* **Mean** = 1.53

* **Median** = 1.00

* **Standard Deviation** = 1.00. This value is low which suggests there isn't much variability within this column (as expected as there are only 4 unique values).

* **Range** = 1 - 4. The range indicates that there are 4 different card types for a customer to hold. 


**PointsEarned:**

* **Mean** = 606.51

* **Median** = 605

* **Standard Deviation** = 225.92. This value is very high which suggest there is significant variability among the amount of points a customer has earned. 

* **Range** = 119 - 1000. The reflects the vast distributions of points earned among the dataset population. 

## Explore Feature Engineering

Feature engineering will be explored in the notebook to identify the best strategies, to be implemented into the ETL pipeline.

Create a new column to group continuous age values for improved interpretability, enhanced analysis and simplified modelling.

Group age into the following bins:

* 18-29
* 30-39
* 40-49
* 50-59
* 60-69
* 70 +

In [15]:
# Group age into bins
# Define the bins and labels for age groups
bins = [ 18, 30, 40, 50, 60, 70, np.inf]
labels = ['18-29', '30-39', '40-49', '50-59', '60-69', '70+']

# Create a new column 'AgeGroup' in the DataFrame (CoPilot Assitance)
df ['AgeGroup'] = pd.cut(df['Age'], bins=bins, labels=labels, right=False)


In [20]:
# Display the first few rows of the DataFrame to verify the new column
df.head()

Unnamed: 0  RowNumber  CustomerId  CreditScore Geography  Gender  Age  \
0           0          1    15598695          619    France  Female   42   
1           1          2    15649354          608     Spain  Female   41   
2           2          3    15737556          502    France  Female   42   
3           3          4    15671610          699    France  Female   39   
4           4          5    15625092          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  Complain  Satisfaction Score Card Type  \
0        101348.88       1         1                   2   DIAMOND   
1        112542.58       0         1                   3   DIAMOND   
2        113931.57       1         1                   3   DIAMOND   
3         93826.63       0         0                   5      GOLD   
4         79084.10       0         0                   5      GOLD   

   Point Earned AgeGroup  
0           464    40-49  
1           456    40-49  
2           377    40-49  
3           350    30-39  
4           425    40-49

_______________

## Outlier Detection